docs/madr/decisions/011-rate-limit-policy.md

```diff
@@ -0,0 +1,294 @@
+     onRateLimit:
+       status: 423
+       headers:
+         - key: "x-Kuma-rate-limited"
+           value: "true"
+           append: true
+```
+We are using [Envoy's local rate limit](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/local_rate_limit_filter) which supports only HTTP configuration.
+ 
+The configuration translates to the Envoy configuration on the route and the listener filter:
+ 
+```yaml
+{
+ "match": {
+   "prefix": "/",
+   "headers": [
+   {
+     "name": "x-kuma-tags",
+     "safe_regex_match": {
+     "google_re2": {},
+     "regex": ".*&kuma.io/service=[^&]*redis_kuma-demo_svc_6379[,&].*"
+     }
+   }
+   ]
+ },
+ "route": {
+   "cluster": "localhost:5000",
+   "timeout": "0s"
+ },
+ "typed_per_filter_config": {
+   "envoy.filters.http.local_ratelimit": {
+   "@type": "type.googleapis.com/envoy.extensions.filters.http.local_ratelimit.v3.LocalRateLimit",
+   ...
+   }
+ }
+}
+ 
+```
+ 
+### Specification
+ 
+The differences between TCP and HTTP rate limiting in `targetRef` and `from` requires different validation of objects.
+Idea is to create one policy but with `oneof` and use different validation methods for them.
+ 
+#### **HTTP Rate Limit**
+ 
+Rate limiting can be configured on both HTTP connection managers and routes. Thanks to the header `x-kuma-tags`, which is propagated and set by Envoy, we can recognize the origin of the HTTP request.
+ 
+#### Top level
+ 
+Top-level targetRef can have all available kinds:
+```yaml
+targetRef:
+ kind: Mesh|MeshSubset|MeshService|MeshServiceSubset|MeshGatewayRoute|MeshHTTPRoute
+ name: ...
+```
+Rate limiting is an inbound policy so only `from` should be configured.
+ 
+#### From level
+ 
+```yaml
+from:
+ - targetRef:
+     kind: Mesh|MeshSubset|MeshService|MeshServiceSubset
+     name: ...
+```
+ 
+Matching on MeshGatewayRoute and MeshHTTPRoute does not make sense (there is any route that a request originates from).
+ 
+#### Configuration
+ 
+```yaml
+ default:
+   http:
+     requests: 5
+     interval: 10s
+     onRateLimit:
+       status: 423
+       headers:
+         - key: "x-kuma-rate-limited"
+           value: "true"
+           append: true
+```
+ 
+#### **TCP Connection Rate Limit**
+ 
+Rate limiting on TCP connection is more complicated because there is no easy way to match only for specific clients.
+Based on alternatives we decided to make it only available for all the clients.
+ 
+#### Top level
+ 
+```yaml
+targetRef:
+ kind: Mesh|MeshSubset|MeshService|MeshServiceSubset
+ name: ...
+```
+Matching on MeshGatewayRoute and MeshHTTPRoute does not make sense because we are configuring TCP listeners.
+Rate limiting is an inbound policy so only `from` should be configured.
+ 
+#### From level
+ 
+```yaml
+from:
+ - targetRef:
+     kind: Mesh
+     name: ...
+```
+ 
+There is no easy way to match requesting services with the specific filter chain that's why we decided to support only configuring TCP rate limiting from all services.
```

Comment:
Is this the recurring issue (#1256) of missing propagation to further filter the service id extracted at the TLS level?

Reply:
Actually seems like #1256 only applies to L7. @jakubdyszkiewicz there's no way at L4 to do this?

Reply:
At L7 it is easier but at L4 I couldn't find a solution.

Reply:
I also don't see a reasonable way to do this on L4.